### PR TITLE
Issue 30285 clean up the unique fields table

### LINKED ISSUE #30285: Clean up the unique_fields table
### Parent Issue

* #29459

### User Story

When a Contentlet with a Unique Field, a Unique Field itself, or a Content Type that contains a Unique Field is deleted, we need to clean up the `unique_fields` table. Here's what to do in each case:

* **Case 1: Deleting a Contentlet with a Unique Field**: Look for the record containing this Contentlet by checking the `contentletIds` property in the `supporting_values` JSON field. If the list contains multiple Contentlets, the specific Contentlet must be removed from the list. If it contains the Contentlet you created, the entire record is deleted.
* **Case 2: Deleting a Unique Field**: Using the field variable attribute in the `supporting_values` JSON field, all records containing such a field must be deleted.
* **Case 3: Deleting a Content Type with at least one Unique Field**: All records having that Content Type ID in the `supporting_values` JSON field must be deleted.

We also need to implement several API/Factory methods to handle these cases.

### Acceptance Criteria

Clean up the `unique_fields` table.

### Steps to Reproduce

* In your `docker-compose.yml` file, set the following env var:
```yml
DOT_ FEATURE_FLAG_DB_UNIQUE_FIELD_VALIDATION: 'true'
```
* Start dotCMS with the Full Starter.
* **THIS IS NOT NECESSARY ANYMORE!** ~At this point, the Upgrade Task that creates the `unique_fields` table is not automatically, so you need to call the class directly. You can use Postman for that:~
```
URL: /v1/upgradetask
Method POST
Body: {
"upgradeTaskClass": "com.dotmarketing.startup.runonce.Task241007CreateUniqueFieldsTable"
}
```
* Log into the PostgreSQL database and verify that (1) the table is there and (2) it has records in it. For the Full Starter, it should be 236 records total.
* Create a Content Type named `My CT` with a single field:
  * Text field `Title`, marked as `Show in List` and  `Unique`.

#### Case 1
* Create a Contentlet of such a type with this value: `my value`.
* Create another Contentlet with the very same value. You must get the following error message:

Screenshot 2024-12-04 at 10 31 39 AM

* Getting the total count of records in the `unique_fields` table must be 237 now, as we just created a Contentlet.
* Delete such a Contentlet and get the table's total count again. It must go down to 236 again.

#### Case 2
* Update the `My CT` to have one more text field named `Description`, required, but NOT unique.
* Create 3 Contentlets of such a type, with different unique values.
* Check the table's total record count. It must be 239.
* Go to the the `My CT` and delete the unique `Title` field. The table's total record count must go down to 236 again.

#### Case 3
* Re-add the `Title` field and, once again, go to the `Content Search` to the unique values for the 3 records again. and set its values for the unique fields again Check the table's total record count. It must be 239 again.
* Go back to the `Content Types` portlet and delete the `My CT`. This will delete all Contentlets from such a type as well.
* After that, the table's total record count must go down to 236 again.

### Proposed Objective

Core Features

### Proposed Priority

Priority 2 - Important

### External Links... Slack Conversations, Support Tickets, Figma Designs, etc.

_No response_

### Assumptions & Initiation Needs

_No response_

### Quality Assurance Notes & Workarounds

_No response_

### Sub-Tasks & Estimates

_No response_

### PR DESCRIPTION
**Pull request by freddyDOTCMS**
### Proposed Changes
* Create a Listener to listen when a Contentlet or a FIeld is deleted and then we can clean the unique_fields table up

https://github.com/dotCMS/core/pull/30798/files#diff-615b3dd949e41009affc1f2921cfe0c6569a1302deec3469d834f6163d7104a8R138

* Create new Method in the UniqueFieldValidationStrategy to clean up the table

https://github.com/dotCMS/core/pull/30798/files#diff-445f8d01aa4de058eaaf883e573d17ef1123b1e74a9a98120ac156b78f4c6522R131

https://github.com/dotCMS/core/pull/30798/files#diff-445f8d01aa4de058eaaf883e573d17ef1123b1e74a9a98120ac156b78f4c6522R162

* Create method to update the unique_fields table data when a Contentlet is publish or unpublish

https://github.com/dotCMS/core/pull/30798/files#diff-445f8d01aa4de058eaaf883e573d17ef1123b1e74a9a98120ac156b78f4c6522R141-R153

* Implemenetd all this new methods in the DBUniqueFieldValidationStrategy

https://github.com/dotCMS/core/pull/30798/files#diff-445f8d01aa4de058eaaf883e573d17ef1123b1e74a9a98120ac156b78f4c6522R141-R153

This PR fixes: #30285

### CODE REVIEW

File: dotCMS/src/main/java/com/dotcms/contenttype/business/uniquefields/extratable/UniqueFieldDataBaseUtil.java [hunk 1]
```diff
@@ -49,12 +47,39 @@ public class UniqueFieldDataBaseUtil {
...
+                "AND supporting_values->>'" + VARIANT_ATTR + "' = ? " +
+                "AND (supporting_values->>'"+ LANGUAGE_ID_ATTR + "')::INTEGER = ? " +
+                "AND (supporting_values->>'" + LIVE_ATTR + "')::BOOLEAN = false";
+
+
+    private final static String GET_UNIQUE_FIELDS_BY_CONTENTLET_AND_LANGUAGE = "SELECT * FROM unique_fields " +
+            "WHERE supporting_values->'" + CONTENTLET_IDS_ATTR + "' @> ?::jsonb AND (supporting_values->>'" + LANGUAGE_ID_ATTR +"')::INTEGER = ?";
+
+    private final static String GET_UNIQUE_FIELDS_BY_CONTENTLET_AND_VARIANT= "SELECT * FROM unique_fields " +
+            "WHERE supporting_values->'" + CONTENTLET_IDS_ATTR + "' @> ?::jsonb AND supporting_values->>'variant' = ?";
```
Comment (jcastro-dotcms) [inline dotCMS/src/main/java/com/dotcms/contenttype/business/uniquefields/extratable/UniqueFieldDataBaseUtil.java:75]: Let's use `VARIANT_ATTR` here too

Comment (freddyDOTCMS) [reply dotCMS/src/main/java/com/dotcms/contenttype/business/uniquefields/extratable/UniqueFieldDataBaseUtil.java]: done https://github.com/dotCMS/core/pull/30798/files#diff-b0aed2eddcc36be2a2e2f46624345a83c8698db238fe4dc228c0fb13a11e344fR74

File: dotCMS/src/main/java/com/dotcms/contenttype/business/uniquefields/UniqueFieldValidationStrategy.java [hunk 2]
```diff
@@ -120,4 +121,46 @@ default void validateField(final Field field) {
...
+        //Default implementation do nothing
+    }
+
+    /**
+     * Method call after un publish a {@link Contentlet} it allow thw {@link UniqueFieldValidationStrategy} do any extra
+     * work that it need it.
+     *
+     * @param versionInfo {@link Contentlet}'s {@link VersionInfo} before un publish
+     */
+    default void afterUnPublish(final VersionInfo versionInfo){
```
Comment (dsilvam) [inline dotCMS/src/main/java/com/dotcms/contenttype/business/uniquefields/UniqueFieldValidationStrategy.java:151]: rename to `afterUnpublish`

File: dotCMS/src/main/java/com/dotcms/contenttype/business/uniquefields/UniqueFieldValidationStrategy.java [hunk 3]
```diff
@@ -120,4 +121,46 @@ default void validateField(final Field field) {
...
+     * work that it need it.
+     *
+     * @param inode Published {@link Contentlet}'s inode
+     */
+    default void afterPublish(final String inode) {
+        //Default implementation do nothing
+    }
+
+    /**
+     * Method call after un publish a {@link Contentlet} it allow thw {@link UniqueFieldValidationStrategy} do any extra
```
Comment (dsilvam) [inline dotCMS/src/main/java/com/dotcms/contenttype/business/uniquefields/UniqueFieldValidationStrategy.java:146]: unpublish*

File: dotCMS/src/main/java/com/dotcms/contenttype/business/uniquefields/UniqueFieldValidationStrategy.java [hunk 4]
```diff
@@ -120,4 +121,46 @@ default void validateField(final Field field) {
...
+     * from the extra table.
+     *
+     * @param contentlet
+     */
+    default void cleanUp(final Contentlet contentlet, final boolean deleteAllVariant) throws DotDataException {
+        //Default implementation do nothing
+    }
+
+    /**
+     * Method call after publish a {@link Contentlet} it allow thw {@link UniqueFieldValidationStrategy} do any extra
```
Comment (dsilvam) [inline dotCMS/src/main/java/com/dotcms/contenttype/business/uniquefields/UniqueFieldValidationStrategy.java:136]: the* (please read what's written)

File: dotCMS/src/main/java/com/dotcms/contenttype/business/uniquefields/UniqueFieldValidationStrategy.java (hunk 3 above)
Comment (dsilvam) [reply dotCMS/src/main/java/com/dotcms/contenttype/business/uniquefields/UniqueFieldValidationStrategy.java:146]: the* (please read what's written)

File: dotCMS/src/main/java/com/dotcms/contenttype/business/uniquefields/UniqueFieldValidationStrategy.java (hunk 2 above)
Comment (dsilvam) [reply dotCMS/src/main/java/com/dotcms/contenttype/business/uniquefields/UniqueFieldValidationStrategy.java:151]: unpublish*

**Review by jcastro-dotcms — APPROVED** [review]

File: dotCMS/src/main/java/com/dotmarketing/business/DeleteContentletVersionInfoEvent.java [hunk 5]
```diff
@@ -0,0 +1,28 @@
+package com.dotmarketing.business;
+
+import com.dotmarketing.portlets.contentlet.model.Contentlet;
+
+import java.io.Serializable;
+
+public class DeleteContentletVersionInfoEvent implements Serializable {
```
Comment (dsilvam) [inline dotCMS/src/main/java/com/dotmarketing/business/DeleteContentletVersionInfoEvent.java:7]: javadoc

File: dotcms-integration/src/test/java/com/dotcms/contenttype/business/FieldAPIImplIntegrationTest.java [hunk 6]
```diff
@@ -168,4 +190,59 @@ public void shouldSaveFields() throws DotDataException, DotSecurityException {
...
+
+    /**
+     * Method to test: {@link FieldAPIImpl#delete(Field, User)}
+     * When: Create a COntentType with a unique fields and later remove the unique Field
+     * Should: Clean the unique_fields extra table
+     *
+     * @throws DotDataException
+     */
+    @Test
+    public void cleanUpUniqueFieldTableAfterDleeteField() throws DotDataException {
```
Comment (dsilvam) [inline dotcms-integration/src/test/java/com/dotcms/contenttype/business/FieldAPIImplIntegrationTest.java:203]: Delete*

File: dotCMS/src/main/java/com/dotmarketing/business/DeleteContentletVersionInfoEvent.java (hunk 5 above)
Comment (freddyDOTCMS) [reply dotCMS/src/main/java/com/dotmarketing/business/DeleteContentletVersionInfoEvent.java:10]: done https://github.com/dotCMS/core/pull/30798/commits/44e8e93dd827f27595fe6f8674e3ba6d598ab341#diff-0ad94e6a4f16335bfb97bb19d51b4742f01ab012e6192293023a6f597fddd51bR7

File: dotCMS/src/main/java/com/dotcms/contenttype/business/uniquefields/UniqueFieldValidationStrategy.java (hunk 4 above)
Comment (freddyDOTCMS) [reply dotCMS/src/main/java/com/dotcms/contenttype/business/uniquefields/UniqueFieldValidationStrategy.java]: done https://github.com/dotCMS/core/pull/30798/commits/44e8e93dd827f27595fe6f8674e3ba6d598ab341#diff-445f8d01aa4de058eaaf883e573d17ef1123b1e74a9a98120ac156b78f4c6522R135

File: dotCMS/src/main/java/com/dotcms/contenttype/business/uniquefields/UniqueFieldValidationStrategy.java (hunk 2 above)
Comment (freddyDOTCMS) [reply dotCMS/src/main/java/com/dotcms/contenttype/business/uniquefields/UniqueFieldValidationStrategy.java]: done https://github.com/dotCMS/core/pull/30798/commits/44e8e93dd827f27595fe6f8674e3ba6d598ab341#diff-445f8d01aa4de058eaaf883e573d17ef1123b1e74a9a98120ac156b78f4c6522R150

File: dotCMS/src/main/java/com/dotcms/contenttype/business/uniquefields/UniqueFieldValidationStrategy.java (hunk 3 above)
Comment (freddyDOTCMS) [reply dotCMS/src/main/java/com/dotcms/contenttype/business/uniquefields/UniqueFieldValidationStrategy.java]: done https://github.com/dotCMS/core/pull/30798/commits/44e8e93dd827f27595fe6f8674e3ba6d598ab341#diff-445f8d01aa4de058eaaf883e573d17ef1123b1e74a9a98120ac156b78f4c6522R135

**Review by dsilvam — APPROVED** [review]

**Pull request merged**